py-rse/correct.Rmd

```diff
@@ -633,471 +469,51 @@ def test_text_must_not_be_empty():
+So for example,
+if we are off by 1 in approximating 8+1 and 56+1,
 we have the same absolute error,
-but the relative error is larger in the first case than in the second.
+but the relative error is larger in the first case than it is in the second.
 
-Relative error is almost always more important than absolute error when we are testing software
-because it makes little sense to say that we're off by a hundredth
-when the value in question is a billionth.
-[Accuracy][accuracy] is how close our answer is to right,
+Relative error is almost always more important than absolute error when we are testing software.
+It makes little sense to say that we're off by a hundredth
+when the value is a billionth.
+[Accuracy][accuracy] is how close our answer is to being right,
 and [precision][precision] is how close repeated measurements are to each other.
 We can be precise without being accurate (systematic bias),
 or accurate without being precise (near the right answer but without many significant digits).
-
-Accuracy is usually more important than precision for human decision making:
-a relative error of 10<sup>-2</sup> (two decimal places) is good enough for most data science
+Accuracy is generally more important than precision for human decision making:
+a relative error of 10^-2^ (two decimal places) is good enough for most data science
 because the decision a human being would make won't change if the number changes by 1%.
 
-We now come to the crux of this lesson:
-if the function we are testing uses floating point numbers,
-what should we compare its result to?
-If we compared the sum of the first few numbers in `vals` to the "right" answer,
-the answer could be `False` even though we're doing nothing wrong.
-If we compared it to a previously calculated result that we had stored somehow,
-the match would be exact.
-
-No one has a good generic answer to this problem
-because its root cause is that we're using approximations,
-and each approximation has to be judged in context.
-So what can you do to test your programs?
-If you are comparing to a saved result,
-and the result was saved at full precision,
-you could use exact equality,
-because there is no reason for the new number to differ.
-However,
-any change to your code,
-however small,
-could trigger a report of a difference.
-Experience shows that these spurious warnings quickly lead developers to stop paying attention to their tests.
-
-A much better approach is to write a test that checks whether numbers are the same within some [tolerance][tolerance],
-which is best expressed as a relative error.
-R's `expect_equal` does this automatically.
 In Python,
 we can use `pytest.approx`,
 which works on lists, sets, arrays, and other collections,
 and can be given either relative or absolute error bounds.
 To show how it works,
 here's an example with an unrealistically tight absolute bound:
 
-```python
+```{python}
 from pytest import approx
```

Comment:
I didn't know about `approx`, I have been using `np.isclose`, guessing `approx` is better for the job but `isclose` has portability so might be worth a mention?

Reply:
Wouldn't `isclose` be *less* portable because it relies on numpy? While approx relies on pytest which we use anyway for unit testing?

Reply:
That's one way to see it! I mean more portable across testing suites, since it is not relying on pytest specifics